src/ssl.c

```diff
@@ -39412,7 +39487,54 @@ void* wolfSSL_SESSION_get_ex_data(const WOLFSSL_SESSION* session, int idx)
 #ifdef HAVE_EX_DATA
     session = ClientSessionToSession(session);
     if (session != NULL) {
-        return wolfSSL_CRYPTO_get_ex_data(&session->ex_data, idx);
+#ifndef NO_SESSION_CACHE
+        if (!session->ownExData) {
+            /* Need to retrieve the data from the session cache */
+            int row;
+            int i;
+            int error = 0;
+            SessionRow* sessRow = NULL;
+            const byte* id;
+            void* ret = NULL;
+            byte foundCache = 0;
+
+            id = session->sessionID;
+            if (session->haveAltSessionID)
+                id = session->altSessionID;
+
+            row = (int)(HashSession(id, ID_LEN, &error) % SESSION_ROWS);
+            if (error != 0) {
+                WOLFSSL_MSG("Hash session failed");
+                return NULL;
+            }
+
+            sessRow = &SessionCache[row];
+            if (SESSION_ROW_LOCK(sessRow) != 0) {
+                WOLFSSL_MSG("Session row lock failed");
+                return NULL;
+            }
+
+            for (i = 0; i < SESSIONS_PER_ROW && i < sessRow->totalCount; i++) {
+                if (XMEMCMP(id, sessRow->Sessions[i].sessionID, ID_LEN) == 0) {
+                    ret = wolfSSL_CRYPTO_get_ex_data(
+                            &sessRow->Sessions[i].ex_data, idx);
+                    foundCache = 1;
+                    break;
+                }
+            }
+            SESSION_ROW_UNLOCK(sessRow);
+            /* If we don't have a session in cache then clear the cache and
+             * own it */
+            if (!foundCache) {
+                WOLFSSL_SESSION* s = (WOLFSSL_SESSION*)session;
+                XMEMSET(&s->ex_data, 0, sizeof(WOLFSSL_CRYPTO_EX_DATA));
+                s->ownExData = 1;
+            }
+            return ret;
+        }
+        else
+#endif
```

Comment:
Can you add braces after this to clarify there is an else above? Helps future-proof code for possible issues down the road.

Reply:
Adding braces.